db/hints: Modularize manager.hh

dawmd opened the pull request:
This PR modularizes `manager.{hh, cc}` by dividing the files into separate smaller units. The changes improve overall readability of code and help reason about it. Each file has a specific purpose now.

This is the first step in refactoring the Hinted Handoff module.

Refs scylladb/scylla#15358

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3616/

piodul commented on db/hints/internal/hint_host_manager.hh:
```diff
@@ -39,7 +39,7 @@ namespace internal {
 class end_point_hints_manager {
 private:
-    friend class sender;
+    friend class hint_sender;
```
> db/hints: Rename sender to hint_sender 

Typo in the commit message: `puporse` -> `purpose`

piodul reviewed (changes requested):
Left some nitpicks, but otherwise LGTM. It is easier to navigate through the code now, which is good.

dawmd commented on db/hints/internal/hint_storage.hh:
```diff
@@ -0,0 +1,31 @@
+namespace db::hints {
+namespace internal {
+
+using node_to_hint_store_factory_type = utils::loading_shared_values<endpoint_id, db::commitlog>;
+using hints_store_ptr = node_to_hint_store_factory_type::entry_ptr;
+using hint_entry_reader = commitlog_entry_reader;
```
Yes. An instance of `node_to_hint_store_factory_type` is part of shard hint manager, and the other two are used in `hint_host_manager.cc` and `hint_sender.cc`. I moved them here because:
* they're related to commitlog,
* `hints_store_ptr` depends on `node_to_hint_store_factory_type`, but they're used in two different files (`manager.hh` vs. `hint_host_manager.cc`).

I think this is a good place to place them.

They should become internal later on when commitlog has finally been isolated. It seems the alises for them in `manager.hh` are unnecessary at the moment, though. I'll remove them.

piodul replied:
Sorry, I got totally confused about the first two and didn't see the dependency between them. As for the last one (`hint_entry_reader`), we will see how the final version of the code looks like. Now, the hint sender is also responsible for the storage because it reads from the commitlog (using the aforementioned hint entry reader). Hopefully, it will make more sense after further refactors.

dawmd commented on db/hints/internal/hint_host_manager.cc:
```diff
@@ -834,5 +834,9 @@ resource_manager& end_point_hints_manager::shard_resource_manager() {
     return _shard_manager._resource_manager;
 }
 
+bool end_point_hints_manager::sender::replay_allowed() const noexcept {
```
I agree the commit title could be better. The premise of these changes is the fact that we cannot have a cycle in `#include`s. Because the declaration of `sender` is moved to a separate header file in the following commit, and `hint_host_manager.hh` includes that header, `sender` has to rely on `end_point_hints_manager` being an incomplete type in its declaration. A consequence of that is that we won't be able to access methods of `end_point_hints_manager` from `hint_sender.hh`. These changes prepare the ground for it, but maybe doing so while moving the declaration of `sender` would've been a better idea because the reason why we do that would be clearer.

Do you think I should reword the commit message or squash the commits?

piodul replied:
Please reword the commit message. The title could just be that you are moving the implementation from header to cc file, and the rest of the message should explain the rationale behind it (you are going to move the declaration of `end_point_hints_manager` to a separate header and don't want to include it; `sender::replay_allowed` needs endpoint hint manager's definition so you need to move it to the cc file, etc.).

dawmd commented on db/hints/internal/hint_host_manager.cc:
```diff
@@ -32,7 +32,7 @@
 namespace internal {
 
-bool end_point_hints_manager::store_hint(schema_ptr s, lw_shared_ptr<const frozen_mutation> fm, tracing::trace_state_ptr tr_state) noexcept {
+bool hint_host_manager::store_hint(schema_ptr s, lw_shared_ptr<const frozen_mutation> fm, tracing::trace_state_ptr tr_state) noexcept {
```
That's a good point. I'll still change the name to be consistent, but `hint_endpoint_manager` should work better.

piodul replied:
Yes, my main concern was about "host" part, adding the "host_" prefix is OK.

dawmd commented:
I updated the PR:
* commit messages have become more compact and corrected
* all `gms::inet_address` instances have been replaced by `endpoint_id`
* the redundant comment has been removed from `hint_logger.hh`
* `hint_storage.hh` has an empty line at the end of the file now
* `hint_host_manager` has become `hint_endpoint_manager`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3690/

denesb reviewed (approved):
LGTM

Although not a problem in this PR, as you continue working on this refactoring, watch out  to not create too many small .cc files. Traditionally we consider that as an anti-pattern because we pay the price of splitting at link time. Maybe this conventional wisdom is outdated, feel free to [experiment](https://github.com/scylladb/scylladb/blob/master/docs/dev/compilation-time-analysis.md) and prove it wrong.
On the other hand we have observed overly large .cc files to make the compile times, especially those on incremental recompilation explode. So neither extreeme is good.
